src/epics.c

```diff
@@ -0,0 +1,581 @@
+  if (ret != ECA_NORMAL) {
+    // FIXME: report error back to start_thread()
+    ERROR("epics plugin: CA error %s occurred while trying to start channel "
+          "access",
+          ca_message(ret));
+    return (void *)1;
+  }
+
+  ca_replace_printf_handler(&printf_handler);
+
+  for (size_t i = 0; i < epics_plugin.pvs_num; ++i) {
+    struct pv *p = &epics_plugin.pvs[i];
+
+    ret = ca_create_channel(p->name, &connection_handler, p, 0, &p->id);
+    if (ret != ECA_NORMAL) {
+      ERROR("epics plugin: CA error %s occurred while trying to create channel "
+            "\"%s\"",
+            ca_message(ret), p->name);
+      ret = 1;
+      goto error;
+    }
+  }
+
+  const double timeout = 2.0;
+  while (epics_plugin.thread_loop != 0) {
+    ca_pend_event(timeout);
+  }
+
+error:
+  for (size_t i = 0; i < epics_plugin.pvs_num; ++i) {
+    struct pv *p = &epics_plugin.pvs[i];
+
+    if (p->eid) {
+      ca_clear_subscription(p->eid);
+    }
+    ca_clear_channel(p->id);
+  }
+
+  ca_context_destroy();
+
+  return (void *)ret;
+}
+
+static int start_thread(void) {
+  pthread_mutex_lock(&epics_plugin.lock);
+
+  int ret = 0;
+  if (epics_plugin.thread_loop != 0) {
+    goto epics_unlock;
+  }
+
+  epics_plugin.thread_loop = 1;
+  ret = plugin_thread_create(&epics_plugin.thread_id, epics_thread, (void *)0,
+                             "epics");
+  if (ret != 0) {
+    epics_plugin.thread_loop = 0;
+    ERROR("epics plugin: Starting thread failed: %d", ret);
+
+    goto epics_unlock;
+  }
+
+  // FIXME: wait untill ca_context_create success
+
+epics_unlock:
+  pthread_mutex_unlock(&epics_plugin.lock);
+
+  return ret;
+}
+
+static int stop_thread(void) {
+  pthread_mutex_lock(&epics_plugin.lock);
+
+  epics_plugin.thread_loop = 0;
+
+  pthread_mutex_unlock(&epics_plugin.lock);
+
+  return pthread_join(epics_plugin.thread_id, NULL);
+}
+
+/*
+<Plugin epics>
+  <Family "cavity.temperature"> # metric family name
+    Type "counter" # or "gauge"
+    Unit "Cel" # for degrees celsius
+    Label "Cavity1:Name" # populate additional metric labels
+    Metric "Cavity1:Temp" # where to read the value from
+  </Family>
+</Plugin>
+*/
+
+static int epics_config_metric(oconfig_item_t *ci, struct pv *p,
+                               size_t mf_index) {
+  if (cf_util_get_string(ci, &p->name) != 0 || p->name == NULL) {
+    ERROR("epics plugin: Wrong metric configuration");
+
+    return -1;
+  }
+
+  p->mf_index = mf_index;
+
+  return 0;
+}
+
+static int epics_config_label(oconfig_item_t *ci, struct pv *p,
+                              size_t mf_index) {
+  if (cf_util_get_string(ci, &p->name) != 0 || p->name == NULL) {
+    ERROR("epics plugin: Wrong label configuration");
+
+    return -1;
+  }
+
+  p->mf_index = mf_index;
+  p->is_label = 1;
+
+  return 0;
+}
+
+static int epics_config_metric_family(oconfig_item_t *ci, size_t mf_index) {
+  metric_family_t *fam = &epics_plugin.mfs[mf_index];
```

Comment:
`mf_index` is required to call `epics_config_metric` and `epics_config_label`. I decided to store family index instead of pointer in order to allow `mfs` reallocations. Otherwise, dangling pointers fix would be required on each `mfs` reallocation.